File: Algo/JavaScript/sumOfLinkedLists.js
class LinkedList {
    constructor(value) {
      this.value = value;
      this.next = null;
    }
  }
  
  function sumOfLinkedLists(linkedListOne, linkedListTwo) {
    let head = new LinkedList(0);
    let currentNode = head;
    let carryOver = 0;
    let valueOne = 0;
    let valueTwo = 0;
    while(linkedListOne || linkedListTwo || carryOver) {
      if (linkedListOne && linkedListTwo) {
        valueOne = linkedListOne.value;
        valueTwo = linkedListTwo.value;
        linkedListOne = linkedListOne.next;
        linkedListTwo = linkedListTwo.next;
      } else if(linkedListOne && !linkedListTwo) {
        valueOne = linkedListOne.value;
        valueTwo = 0;
        linkedListOne = linkedListOne.next;
      } else if(!linkedListOne && linkedListTwo) {
        valueOne = 0;
        valueTwo = linkedListTwo.value;
        linkedListTwo = linkedListTwo.next;
      } else if(!linkedListOne && !linkedListTwo && carryOver) {
        valueOne = 0;
        valueTwo = 0;
      }
  
      const total = valueOne + valueTwo + carryOver;
      const sum = total % 10;
  
      if (!linkedListOne && !linkedListTwo && total < 10) {
        carryOver = null;
      } else {
        carryOver = Math.floor(total / 10);
      }
  
      let nextNode = new LinkedList(sum);
      currentNode.next = nextNode;
      currentNode = nextNode;
    }
    const listOfSums = head.next;
    
    return listOfSums;
  }

File: Algo/Ruby/sum_of_linked_lists.rb
class LinkedList
    attr_accessor :value, :next
    def initialize(value)
        @value = value
    end

    def self.sum_of_linked_lists(linkedListOne, linkedListTwo)
        current_node_one = linkedListOne
        current_node_two = linkedListTwo
        carry_over = 0
        head = LinkedList.new(0)
        current_node = head
        
        while(current_node_one || current_node_two || carry_over)
           if current_node_one && current_node_two
            value_one = current_node_one.value
            value_two = current_node_two.value
            current_node_one = current_node_one.next
            current_node_two = current_node_two.next
           elsif current_node_one.nil? && current_node_two
            value_one = 0
            value_two = current_node_two.value
            current_node_two = current_node_two.next
           elsif current_node_one && current_node_two.nil?
            value_one = current_node_one.value
            value_two = 0
            current_node_one = current_node_one.next
           elsif current_node_one.nil? && current_node_two.nil? && carry_over
            value_one = 0
            value_two = 0
           end
            total = value_one + value_two + carry_over
            sum = total % 10
            
            if current_node_one.nil? && current_node_two.nil? && total < 10
                carry_over = nil
            else
                carry_over = total/10
            end
            
            next_node = LinkedList.new(sum)
            current_node.next = next_node
            current_node = next_node
        end

        head.next
    end
end

linked_list_one = LinkedList.new(2)
node2 = LinkedList.new(4)
node3 = LinkedList.new(7)
node4 = LinkedList.new(1)

linked_list_one.next = node2
node2.next = node3
node3.next = node4

linked_list_two = LinkedList.new(9)
node5 = LinkedList.new(4)
node6 = LinkedList.new(5)

linked_list_two.next = node5
node5.next = node6

p LinkedList.sum_of_linked_lists(linked_list_one, linked_list_two)
